docs/src/tutorials/nonlinear/user_defined_hessians.jl

```diff
@@ -0,0 +1,264 @@
+    return objective_value(model), value.(y)
+end
+
+# The next function takes a value of ``x`` and returns the optimal lower-level
+# objective-value and the optimal response ``y``. The reason why we need both
+# the objective and the optimal ``y`` will be made clear shortly, but for now
+# let us define:
+
+function V(x...)
+    f, _ = solve_lower_level(x...)
+    return f
+end
+
+# Then, we can substitute ``V`` into our full problem to create:
+
+# ```math
+# \begin{array}{r l}
+# \min\limits_{x} & x_1^2 + x_2^2 + V(x_1, x_2) \\
+# s.t.            & x \ge 0.
+# \end{array}
+# ```
+
+# This looks like a nonlinear optimization problem with a user-defined function
+# ``V``! However, because ``V`` solves an optimization problem internally, we
+# can't use automatic differentiation to compute the first and second
+# derivatives. Instead, we can use JuMP's ability to pass callback functions
+# for the gradient and Hessian instead.
+
+# First up, we need to define the gradient of ``V`` with respect to ``x``. In
+# general, this may be difficult to compute, but because ``x`` appears only in
+# the objective, we can just differentiate the objective function with respect
+# to ``x``, giving:
+
+function ∇V(g::AbstractVector, x...)
+    _, y = solve_lower_level(x...)
+    g[1] = 2 * x[1] * y[1] - y[1]^4
+    g[2] = 2 * x[2] * y[2] - 2 * y[2]^4
+    return
+end
+
+# Second, we need to define the Hessian of ``V`` with respect to ``x``. This is
+# a symmetric matrix, but in our example only the diagonal elements are
+# non-zero:
+
+function ∇²V(H::AbstractMatrix, x...)
+    _, y = solve_lower_level(x...)
+    H[1, 1] = 2 * y[1]
+    H[2, 2] = 2 * y[2]
+    return
+end
+
+# We now have enough to define our bilevel optimization problem:
+
+model = Model(Ipopt.Optimizer)
+@variable(model, x[1:2] >= 0)
+register(model, :V, 2, V, ∇V, ∇²V)
+@NLobjective(model, Min, x[1]^2 + x[2]^2 + V(x[1], x[2]))
+optimize!(model)
+solution_summary(model)
+
+# The optimal objective value is:
+
+objective_value(model)
+
+# and the optimal upper-level decision variables ``x`` are:
+
+value.(x)
+
+# To compute the optimal lower-level decision variables, we need to call
+# `solve_lower_level` with the optimal upper-level decision variables:
+
+_, y = solve_lower_level(value.(x)...)
+y
+
+# ## Memoization
+
+# Our solution approach works, but it has a performance problem: every time
+# we need to compute the value, gradient, or Hessian of ``V``, we have to
+# re-solve the lower-level optimization problem! This is wasteful, because we
+# will often call the gradient and Hessian at the same point, and so solving the
+# problem twice with the same input repeats work unnecessarily.
+
+# We can work around this by using [memoization](https://en.wikipedia.org/wiki/Memoization).
+# First, we create a struct to cache the results:
+
+mutable struct Cache
+    x::Any
+    f::Float64
+    y::Vector{Float64}
+end
+
+cache = Cache(Float64[], NaN, Float64[])
```

Comment:
Untyped global variables :(
We don't want people following this example.
How about
```jl
function cached_f(cache, x...)
   ...
end
...
cache = Cache(...)
register(model, :V, 2, (x...) -> cached_f(cache, x...), ...)
```

Reply:
I considered adding a `const`, but this would also work.

Reply:
The global var version is very limiting, I'd rather push users in a better direction.